feat: shielded TX support

karim-en opened the pull request:
## What This Does

Adds security validation for Orchard (shielded) withdrawals in the Zcash bridge. Users can now withdraw to Unified Addresses with proper enforcement that bundles go to the right recipient with the right amount.

## The Problem

The `orchard_tx` branch added basic Orchard transaction support, but had TODOs around validation:
- No verification that Orchard outputs go to the intended recipient
- No verification that amounts match what the user is withdrawing
- No protection against bundle tampering after transparent signatures are collected

This meant a malicious user could potentially redirect shielded funds or manipulate amounts.

## The Solution

### 1. OVK-Based Output Recovery

Implemented `orchard_policy.rs` module that uses a hardcoded Bridge OVK to recover and validate Orchard outputs:

```rust
pub const BRIDGE_OVK: [u8; 32] = [0u8; 32];

pub fn validate_orchard_bundle(
    bundle: &Bundle<Authorized, ZatBalance>,
    expected_recipient: &str,
    expected_amount: u64,
    chain: &Chain,
) {
    // Recover output using Bridge OVK
    let (recovered_amount, recovered_addr) = recover_orchard_output(bundle);

    // Validate matches expectations
    require!(recovered_amount == expected_amount, "Orchard amount mismatch");
    require!(recovered_addr == expected_addr_bytes, "Orchard recipient mismatch");
}
```

**Trade-off:** The OVK is public (contract code is public), so anyone can recover these outputs. This is acceptable for the bridge use case - we're enforcing policy, not providing privacy.

### 2. ZIP-244 Sighash Binding

The Orchard bundle is now included in the `TransactionData` used for transparent signature generation. This means:
- Any modification to the Orchard bytes invalidates all transparent signatures
- Prevents post-signing bundle swap attacks


### 3. Orchard-Only Withdrawals

Added support for withdrawals with zero transparent outputs (all funds → shielded pool):

```rust
if psbt.get_output_num() == 0 {
    // Orchard-only case: skip transparent validation
    let gas_fee = max_gas_fee.unwrap_or(10000);
    let orchard_amount = amount - withdraw_fee - gas_fee;
    (orchard_amount, gas_fee)
} else {
    // Normal transparent output validation
    self.check_withdraw_psbt_valid(...)
}
```

### 4. Unified Address Validation

Enforces that Unified Addresses must come with Orchard bundles:

```rust
if target_btc_address.starts_with("u1") || target_btc_address.starts_with("u") {
    require!(orchard_bundle.is_some(),
        "Unified Address provided without Orchard bundle. \
         Either provide both or use a transparent Zcash address (starts with t1/t3)");
}
```

## Test Coverage

Added 5 comprehensive tests covering:
- Happy path e2e withdrawal with OVK validation
- Amount mismatch rejection
- Recipient mismatch rejection (different spending keys → different recipients)
- Unified Address without bundle rejection
- Bundle serialization in final transaction

**Note:** Tests cache generated Orchard bundles because proof generation takes ~80-90s. First run is slow, subsequent runs are fast (~16s total).

```bash
# Build first
cargo near build non-reproducible-wasm --features zcash --out-dir res --no-abi

# Run tests
cargo test -p satoshi-bridge --features zcash \
  --test test_orchard_validation \
  --test test_orchard_withdrawal \
  -- --test-threads=1 --nocapture
```

## Breaking Changes

### Event Encoding
Changed `SignedBtcTransaction` event from hex to base64 to stay under NEAR's 16KB log limit:

```diff
- tx_bytes: &'a Vec<u8>
+ tx_bytes_base64: String
```

Overhead reduced from 2x (hex) to 1.33x (base64). Relayer will need to update the parser.

olga24912 committed:
provide metadata for nBTC

olga24912 committed:
fix satoshi bridge cargo toml

olga24912 committed:
address -> script_pubkey

olga24912 committed:
btc-p2wpkh -> utxo_chain

olga24912 committed:
remove btc_network

olga24912 committed:
draft zcash tx impl

olga24912 committed:
fix consenses decode

olga24912 committed:
zcash_primitives

olga24912 committed:
fix cargo toml

olga24912 committed:
test for script

olga24912 committed:
fix get output

olga24912 committed:
change signhash

karim-en committed:
Refactor transaction type

karim-en committed:
Bump zcash_primitives version

karim-en committed:
Convert btc tx into zcash tx

karim-en committed:
Set `final_script_sig`

karim-en committed:
Fix sighash

karim-en committed:
fix

olga24912 committed:
fix sequence

olga24912 committed:
zcash build

olga24912 committed:
add zcash features

olga24912 committed:
AddressInner -> Address

olga24912 committed:
new line

olga24912 committed:
test parse address

olga24912 committed:
test from pubkey

olga24912 committed:
improve tests

olga24912 committed:
fix warnings

olga24912 committed:
check zcash fee

olga24912 committed:
safe expiry height

olga24912 committed:
fix expiry_height

olga24912 committed:
add rbf

olga24912 committed:
fix sequence

olga24912 committed:
extract expire_height for cancel_withdraw

olga24912 committed:
fix cancel active utxo managment

olga24912 committed:
fix active utxo managment

olga24912 committed:
utxo_mamagment extract block height

olga24912 committed:
gap into config

karim-en committed:
Replace last_block_header with last_block_height

olga24912 committed:
construct output for zcash

olga24912 committed:
add simple try from address

olga24912 committed:
parse unified address

olga24912 committed:
script pubkey

olga24912 committed:
fmt for unified address

olga24912 committed:
fix find receiver in list

olga24912 committed:
fix warnings

olga24912 committed:
safe psbt wraper

olga24912 committed:
deserialize

olga24912 committed:
move het hash to sign

olga24912 committed:
extract tx byte

olga24912 committed:
save signature

olga24912 committed:
incapsulate psbt

olga24912 committed:
separate transaction

olga24912 committed:
separate expiry hight

olga24912 committed:
separate psbt creation

olga24912 committed:
separate chain specific functions

olga24912 committed:
separate zcash functions

olga24912 committed:
separate zcash

olga24912 committed:
fix btc

olga24912 committed:
fix warnings

olga24912 committed:
optimize get min fee

olga24912 committed:
remove psbt for zcash

olga24912 committed:
remove unsed function

olga24912 committed:
add get_input_num and get_output_num

olga24912 committed:
fix satoshi bridge

olga24912 committed:
fix new

olga24912 committed:
make psbt field private

olga24912 committed:
reduce copy past

olga24912 committed:
reduce copypast

olga24912 committed:
fix test contract config for zcash

olga24912 committed:
don't panic

olga24912 committed:
fmt

olga24912 committed:
reduce include

olga24912 committed:
updaye zcash_protocol

olga24912 committed:
fix chain id

olga24912 committed:
remove default

olga24912 committed:
add version

olga24912 committed:
fix bitcoin connector

olga24912 committed:
remove bytes_to_btc_transaction

olga24912 committed:
don't use fully-qualified name

olga24912 committed:
fix bitcoin tests

olga24912 committed:
fix check block heigth for branch id

olga24912 committed:
save block height

olga24912 committed:
fix

olga24912 committed:
fix make

olga24912 committed:
fix valid confiramtions range

karim-en committed:
Bump version

frolvanya committed:
fix: use `U128` type for amount in `ft_on_transfer_callback`

olga24912 committed:
provide max gas fee

olga24912 committed:
Merge branch 'omni-main' into zcash_support

karim-en committed:
Add orchard support

karim-en committed:
Add comments

karim-en committed:
Add comment

olga24912 commented on contracts/satoshi-bridge/src/api/bridge.rs:
```diff
@@ -207,10 +207,15 @@ impl Contract {
+        &mut self,
+        input: Vec<OutPoint>,
+        output: Vec<TxOut>,
+        orchard_bundle: Option<Vec<u8>>,
```
And for what reasons might we need orchard_bundle in active UTXO management? I suggest removing it from here.

olga24912 committed:
fix outputs checks

olga24912 commented on contracts/satoshi-bridge/src/psbt.rs:
```diff
@@ -185,7 +185,7 @@ impl Contract {
+            actual_received_amounts.len() <= 1,
             "only one user output is allowed."
         );
         let actual_received_amount = actual_received_amounts[0];
```
Next come the checks of the amount the user will receive. In shielded transactions we don’t know this amount and don’t know how much was spent on gas.

daira commented on contracts/satoshi-bridge/src/api/token_receiver.rs:
```diff
@@ -8,11 +8,13 @@ pub const GAS_FOR_FT_ON_TRANSFER_CALL_BACK: Gas = Gas::from_tgas(100);
         input: Vec<OutPoint>,
         output: Vec<TxOut>,
         max_gas_fee: Option<U128>,
+        orchard_bundle_bytes: Option<String>,
```
The naming of this is confusing. It's a String containing the hex bytes of the Orchard section of the transaction. This should probably be documented I think (if it's what you intend). We normally use "bundle" to refer to the in-memory `OrchardBundle` structure — which is why at first I posted a comment about memory safety, misunderstanding it a hex encoding of that structure.

daira commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -61,12 +64,28 @@ impl PsbtWrapper {
+        // let value = "<Amount of the output>";
+
+        // TODO: verify orchard bundle
+        // How to verify orchard bundle value and recipient?
```
You can't verify that just based on the bytes of the Orchard section of the transaction. You need to have the data that was encrypted and encoded to produce those bytes. That is available in the PCZT. I think @str4d will need to explain this.

karim-en replied:
@daira, we can't create the zkproof on-chain over PCZT, it is very gas-intensive. We planned to generate it off-chain and then pass it and verify it.
cc @str4d

r-near commented:
@daira Is it possible to avoid on-chain Halo2 proof verification entirely?

What if we
1. Bind transparent signatures to the exact Orchard bytes (include the bundle in ZIP-244 txid parts).
2. Recover the output with a known OVK and check recipient and amount (policy check).

@str4d can you check me here - is my logic correct?

The contract only burns nZEC after on‑chain inclusion, meaning the network has already validated the Orchard proof/signatures. On-chain proof verification seems redundant then, right?

Pre‑inclusion, on-chain proof verification would merely pre‑filter invalid bundles that off‑chain relayers should already reject.

@karim-en this would obviate the need for Halo2 proofs, forking the crates to enable serde, embedding the VK/doing the keygen on-chain.

karim-en committed:
Calculate sighash with `orchard_bundle`

r-near committed:
Implement Orchard Bundle Validation for Zcash Withdrawals (#20)

* feat: zcash validation

* fix: use ZIP-317 formula with orchard action count for fee calculation

Previously, get_min_fee() hardcoded 0 for orchard_action_count when
calling the ZIP-317 fee_required function. This resulted in incorrect
fee calculations for transactions with Orchard bundles.

Now properly counts actions from the orchard bundle (if present) and
passes this to the ZIP-317 FeeRule. The formula accounts for:
- Transparent inputs (P2PKH standard size)
- Transparent outputs (serialized size)
- Orchard actions (actual count from bundle)

The fee calculation varies with transaction structure per ZIP-317
specification - 10,000 satoshis is just the minimum for simple
transactions with minimal inputs/outputs.

* fix: replace hardcoded gas fee with ZIP-317 calculation

Previously used hardcoded 10000 satoshis as default gas fee. Now:

1. Calculate actual fee using ZIP-317 formula via psbt.get_min_fee()
2. Use max_gas_fee as an upper bound when provided
3. Fall back to computed fee when max_gas_fee not provided

This properly accounts for transaction structure (inputs, outputs,
orchard actions) per ZIP-317 spec. The fee varies with tx complexity
rather than using a fixed minimum value.

Changes in contract_methods.rs:
- Create preliminary PSBT to compute fee before validation
- Recreate PSBT with correct expected amounts for validation

Changes in token_receiver.rs:
- Compute ZIP-317 fee and use max_gas_fee as cap

* fix: update minimum orchard actions check from 1 to 2

Per Orchard protocol specification, bundles must contain at least 2
actions for privacy (MIN_ACTIONS = 2). The Orchard builder
automatically pads bundles with dummy actions to meet this minimum.

Changes:
- Added MIN_ACTIONS constant (2) in orchard_policy.rs
- Updated validation to require >= MIN_ACTIONS instead of == 1
- Applied to both strict validation (with expected values) and
  looser validation (internal operations)
- Improved error messages to show expected vs actual action count

Reference: https://github.com/zcash/orchard/blob/main/src/builder.rs#L36

* fix: use base64 encoding only for ZCash to preserve Bitcoin compatibility

Previously, all chains used base64 encoding for tx_bytes in the
SignedBtcTransaction event, which breaks backward compatibility
for Bitcoin indexers expecting hex encoding.

Now:
- ZCash chains (mainnet/testnet): Use base64 (1.33x overhead)
  - Beneficial for larger Orchard transactions
- Bitcoin/other chains: Use hex encoding (backward compatible)

The event field name remains `tx_bytes_base64` but contains hex
for non-ZCash chains to avoid breaking the event structure.

* fix: validate Unified Address for actual Orchard receiver presence

Previously, the code checked if an address starts with "u1" or "u"
to determine if an Orchard bundle is required. This is incorrect
because Unified Addresses can contain only transparent receivers
(P2PKH/P2SH), which is a valid scenario without an Orchard bundle.

Now:
- Added has_orchard_receiver() helper to check if a Unified Address
  actually contains an Orchard receiver
- Only require Orchard bundle if the UA has an Orchard receiver
- Allow Unified Addresses with only transparent receivers to work
  without an Orchard bundle

Transparent addresses can be extracted from Unified Addresses,
which is a valid use case.

* fix: reject mixed Orchard and transparent output scenarios

Previously, the code would crash when a Unified Address with only
an Orchard receiver was provided along with transparent change outputs.
The transparent validation path would fail trying to extract a
transparent script_pubkey from an Orchard-only UA.

Now:
- Added has_orchard_bundle() helper to PsbtWrapper
- Explicitly reject transactions with both Orchard bundle and
  transparent outputs
- Require either pure Orchard (no transparent outputs) or pure
  transparent (no Orchard bundle)

This prevents the classic case where target has Orchard receiver
but change goes to transparent output, which would cause the
program to crash.

The restriction simplifies validation logic and avoids edge cases.
Users must choose one path or the other for clarity.

* fix: validate actual Orchard output amount from bundle

Previously, for Orchard-only withdrawals, the code calculated what the output amount *should* be but didn't validate the actual amount in the Orchard bundle. This was a critical security issue - users could provide any amount in their Orchard bundle.

Now:
- Added get_orchard_output_amount() to PsbtWrapper that recovers the actual output amount using the bridge OVK
- Validate that the recovered Orchard amount is within the expected range (accounting for fees and potential dust adjustment)
- Use the same validation logic as transparent outputs: allow a range between (expected - min_change_amount) and expected

The solution deserializes and validates the Orchard output amount, ensuring users receive the correct amount after fees.

* fix: support mixed Orchard output with transparent change (baseline case)

Previously, commit 9ce848d3001496bbae293fb2ad54c03848f9d1b7 rejected mixed scenarios (Orchard + transparent),
but this was incorrect. The baseline scenario for Orchard withdrawals is:
- 1 Orchard output to the user's Unified Address
- 1+ transparent change outputs back to the bridge

Now properly supports mixed scenarios:
- Validate the Orchard output amount (to user)
- Validate all transparent outputs are change to bridge address
- Ensure change amounts are within min/max limits
- Verify change + gas_fee doesn't exceed input

The new validation logic:
- Orchard bundle present: Validate Orchard output + transparent change
- No Orchard bundle: Use existing transparent-only validation

* fix: comprehensive Orchard bundle validation improvements

Fixed 4 critical validation gaps identified during code review:

1. **Value Balance Validation** (NEW)
   - Now validates bundle.value_balance() equals -(output_amount)
   - Ensures value flows correctly from transparent to Orchard pool
   - Prevents malicious bundles with incorrect value balance

2. **All Outputs Validated** (NEW)
   - Iterates through ALL actions/outputs (not just first)
   - Ensures exactly 1 non-zero output exists
   - Prevents hiding extra outputs in unused actions

3. **Range Validation for Dust Adjustment** (FIXED)
   - Changed from exact match to range validation
   - Allows user to adjust amount for dust (like transparent does)
   - Range: [expected - min_change_amount, expected]
   - Provides same flexibility as transparent withdrawals

4. **Exact Transparent Accounting** (FIXED)
   - Changed `<=` to `==` in transparent change validation
   - Validates: input == orchard_amount + change + gas_fee
   - Prevents missing value in accounting
   - Ensures value balance integrity

These fixes provide defense-in-depth. While the ZCash network would
catch most issues, we now validate comprehensively at contract level
to reject bad requests early and ensure robust accounting.

* tmp contract fix

* Min Actions = 1

* safe output

* remove unused function

* separate validation

* merge checks

* fix add extra amount

* remove unused function

* remove changes

* fmt

* remove amount checks

* remove no needed check

* move orchard validation into check function

* remove gas calculation

* provide expire height

* fix get_branch_id

* reduce copypast

* event beckword compatibility

* fix warnings

* clippy

* fix compilation for bitcoin

* fmt

* fix: add expiry_height field to tests

Add missing expiry_height: None to TokenReceiverMessage::Withdraw
in test files after the field was added to the struct.

* save target_address

* orchard bundle in RBF

* chain specific data

* String -> Base64

* remove dublicate code

* return Error on Orchard Bundle

* check the expected actions num

* clean up

* save recipient_address to psbt

* check version

* fix tests compilation

* fix: update fixtures, exports, and tests

* fix: replace unwrap() with descriptive error messages

- bitcoin_utils/psbt_wrapper: add ERR_INVALID_PSBT_HEX, ERR_EXTRACT_TX,
  ERR_MISSING_WITNESS_UTXO, ERR_SIGHASH messages
- zcash_utils/psbt_wrapper: add ERR_INVALID_ORCHARD_BUNDLE,
  ERR_ORCHARD_VALIDATION, and detailed ERR_INVALID_PSBT messages for
  all deserialization steps
- config: add ERR_CONFIG messages for uninitialized contract/config
- account: include account_id in ERR_ACCOUNT_NOT_REGISTERED message

* provide chain_specific data

* remove unnecessary .clone() calls

* add type alias for Orchard raw address bytes and simplify extract_orchard_bundle

* replace expect() with env::panic_str() to reduce contract size

* Update contracts/satoshi-bridge/src/network.rs

Co-authored-by: Ivan Frolov <59515280+frolvanya@users.noreply.github.com>

* chore: imported `OrchardRawAddress`

---------

Co-authored-by: Olga Kunyavskaya <olga.kunyavskaya@aurora.dev>
Co-authored-by: Ivan Frolov <59515280+frolvanya@users.noreply.github.com>
Co-authored-by: Ivan Frolov <frolvanya@gmail.com>

karim-en commented on contracts/satoshi-bridge/src/account.rs:
```diff
@@ -86,15 +86,15 @@ impl Contract {
             .get(account_id)
             .map(|o| o.into())
-            .expect("ACCOUNT NOT REGISTERED")
+            .unwrap_or_else(|| panic!("ERR_ACCOUNT_NOT_REGISTERED: {}", account_id))
```
The panic! should be replaced with `env::panic_str`

karim-en replied:
0be10b6169ae22baa4486dce47718f19e77e1c58

karim-en commented on contracts/satoshi-bridge/src/account.rs:
```diff
@@ -86,15 +86,15 @@ impl Contract {
             .get_mut(account_id)
             .map(|o| o.into())
-            .expect("ACCOUNT NOT REGISTERED")
+            .unwrap_or_else(|| panic!("ERR_ACCOUNT_NOT_REGISTERED: {}", account_id))
```
The panic! should be replaced with `env::panic_str`

karim-en replied:
0be10b6169ae22baa4486dce47718f19e77e1c58

karim-en commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -239,10 +368,10 @@ impl PsbtWrapper {
             None,
             None,
-            None,
+            self.orchard_bundle.clone(),
```
This can bee optimized by replacing the `to_zcash_tx` with raw `digest` calls

karim-en replied:
https://github.com/Near-One/btc-bridge/pull/6/commits/3612052114c42e3fcf59f2d7f63b37e452f02bed

karim-en replied:
cc @olga24912

karim-en committed:
Avoid extra clone of Orchard bundle

karim-en committed:
Replace panic with panic_str

karim-en committed:
Bump version

karim-en committed:
Fix clippy

karim-en committed:
Fix build variant

karim-en committed:
Fix orchard tests

karim-en committed:
Run lint and tests to CI

karim-en committed:
Merge branch 'omni-main' into orchard_tx

karim-en committed:
Revert default

karim-en committed:
Remove added ci

karim-en committed:
Remove unused fmt

karim-en committed:
Fix tests

karim-en committed:
Fix orchard test

karim-en committed:
Fix orchard test

karim-en committed:
Add missed `safe_deposit`

olga24912 committed:
add different wasm path

olga24912 committed:
fix orchard tests

olga24912 committed:
fix test_base

olga24912 committed:
fix check predecessor account id

olga24912 committed:
fix test

olga24912 commented on contracts/satoshi-bridge/src/psbt.rs:
```diff
@@ -175,32 +175,42 @@ impl Contract {
+
         require!(
-            actual_received_amounts.len() == 1,
+            actual_received_amounts.len() <= 1,
```
In which cases can it be 0?

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/6/commits/c43a72ca8192d840355a19d6442bbbd2a41b11e3

olga24912 commented on contracts/satoshi-bridge/src/rbf/withdraw.rs:
```diff
@@ -74,3 +71,34 @@ impl Contract {
+            target_address_script_pubkey.as_script(),
+            self.internal_config().chain.clone(),
+        )
+        .unwrap()
```
expect

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/6/commits/629a18963ea0be0d8d5c39f3a50d97cbe067e095

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -93,22 +133,50 @@ impl Contract {
+                last_block_height + self.get_config().expiry_height_gap,
+                last_block_height + 2 * self.get_config().expiry_height_gap
+            )
+        );
```
Extract a method to check expiry_height

olga24912 replied:
fixed:  https://github.com/Near-One/btc-bridge/pull/6/commits/dd4487ceaf7627f479457cefd869c9bb1eb0af14

olga24912 committed:
fix check outputs number

olga24912 committed:
unwrap -> expect

olga24912 committed:
extract expiry height check
